README.md
mangos-zero server [![Build Status][build-status-img]][build-status] [![Coverity Build Status][scan-status-img]][scan-status]
==================

**mangos-zero** is an Open Source [GPL version 2 licensed](License.md) server for
running your own instance of the [World of Warcraft][wow], specifically of
[vanilla WoW][wow-1] aka. patch 1.12.

The ultimate goal of the project is to preserve the game in its' first released
version and allow players to relive the experience of the original game, since
the game is always evolving and being updated with new content.

If you are nostalgic or just missed out on the early days of World of Warcraft,
stay with us, as [vanilla WoW][wow-1] is right here.

Supporting the project
----------------------
If you like **mangos-zero** and want to support the project, there are various
options:

* [![Tip the developer!][gittip-img]][gittip] financial support: I spend roughly
  30 hours per week to work on the project, including server, scripts, database
  and documentation. A little [tip][gittip] helps pay the bills.
* testing and reporting issues: if you use **mangos-zero**, please report issues
  you find for the [server][mz-server-issues], [scripts][mz-scripts-issues],
  [database][mz-content-issues] or our [documentation][mz-docs-issues].
* submit fixes: any help is welcome! It does not matter if its' proof reading
  and fixing messages printed by the server, or if you produce documentation
  or code. As long as you have fun, be an active part of the project!

Documentation
-------------
Documentation is available at [docs.getmangos.com][mangos-zero-docs], and is
updated on regular basis with the latest changes. Documentation is available
in [source format][mz-docs], too, if you want to use it locally or contribute.

Installation
------------
To install **mangos-zero** on various target platforms, please checkout our
updated [installation documentation][mangos-zero-installation].

Contributing
------------
All information about contributing to **mangos-zero** can be found in our
[documentation for contributors][mangos-zero-contribute].

Bug Tracking
------------
**mangos-zero** uses the [BitBucket issue tracker][mz-server-issues]. If you
have found bugs, please create an issue.

License
-------
**mangos-zero** is licensed under the terms of the [GNU GPL version 2](License.md),
with a special exception allowing you to link our server against [OpenSSL][openssl].

We provide a few libraries within the `dep` directory for which different
licenses apply.

[wow]:                      http://worldofwarcraft.com/ "World of Warcraft"
[wow-1]:                    http://blizzard.com/games/wow/ "Blizzard Entertainment: World of Warcraft"

[mangos-zero]:              http://getmangos.com/ "Vanilla WoW - getmangos.com"
[mangos-zero-docs]:         http://docs.getmangos.com/ "mangos-zero documentation"
[mangos-zero-contribute]:   http://docs.getmangos.com/en/latest/contributing/ "Contributing - mangos-zero"
[mangos-zero-installation]: http://docs.getmangos.com/en/latest/installation/ "Installation - mangos-zero"

[mz-team]:                  http://bitbucket.org/mangoszero/ "mangos-zero project"
[mz-server]:                http://bitbucket.org/mangoszero/server/ "mangos-zero server repository"
[mz-scripts]:               http://bitbucket.org/mangoszero/scripts/ "mangos-zero scripts repository"
[mz-content]:               http://bitbucket.org/mangoszero/content/ "mangos-zero database repository"
[mz-docs]:                  http://bitbucket.org/mangoszero/documentation/ "mangos-zero documentation repository"

[mz-server-issues]:         http://bitbucket.org/mangoszero/server/issues "mangos-zero server issues"
[mz-scripts-issues]:        http://bitbucket.org/mangoszero/scripts/issues "mangos-zero scripts issues"
[mz-content-issues]:        http://bitbucket.org/mangoszero/content/issues "mangos-zero database issues"
[mz-docs-issues]:           http://bitbucket.org/mangoszero/documentation/issues "mangos-zero documentation issues"

[openssl]:                  http://openssl.org/ "OpenSSL: The Open Source toolkit for SSL/TLS"

[build-status]:             http://drone.io/bitbucket.org/mangoszero/server/latest "mangos-zero build status"
[build-status-img]:         http://drone.io/bitbucket.org/mangoszero/server/status.png "mangos-zero build status image"

[scan-status]:              http://scan.coverity.com/projects/365 "mangos-zero scan status"
[scan-status-img]:          http://scan.coverity.com/projects/365/badge.svg "mangos-zero scan status image"

[gittip]:                   http://gratipay.com/danielsreichenbach/
[gittip-img]:               http://img.shields.io/gratipay/danielsreichenbach.svg


ChangeLog.md
Change log for mangos-zero
==========================

This change log references the relevant changes (bug and security fixes) done
in 1.0 minor versions.

## 1.2.0 (2014-12-XX)

Major changes for this build which require your attention when upgrading include
things such as these:

*

Minor fixes and improvements have been added, such as:

*

## 1.1.0 (2014-12-03)

Major changes for this build which require your attention when upgrading include
things such as these:

* support for ScriptDev2 has been dropped, and the matching configuration options
  are gone. Instead we provide our own script library, where we recreate scripts
  for game content from scratch (use WITH_MZ_SCRIPTS parameter to enable/disable).
* fixed resting system. Resting will now work as intended.
* SOAP bindings removed! In order to focus our efforts on the server and increasing
  its' quality, will continue to strip down the server to relevant components.
* remote access removed! To reduce possible security risk, we stripped the remote
  access console from the world server.
* **new** configuration setting `LoadAllGridsOnMaps`: with this setting a list of
  maps can be defined on which grids and creatures will be kept loaded. Continents
  will be loaded on server start, instances when a player enters. **Notice** that
  this will increase CPU load and memory usage depending on the amount of maps you
  chose to keep in memory.
* load active entities based on active maps. Ideas and loop to load all grids on
  a map by Neo2003.
* implement `CREATURE_FLAG_EXTRA_ACTIVE`: setting this flag in `creature_template`
  `ExtraFlags` will result in the creature being loaded and set to active when the
  map is created. Use with care.
* added a new admin command to force raid instance resets
* implement new TEMPFACTION_TOGGLE flags for `UNIT_FLAG_PACIFIED` and
  `UNIT_FLAG_NOT_SELECTABLE`
* implement TARGET_UNIT_NEARBY_ALLY and unify handling of TARGET_UNIT_NEARBY_
* allow target 60 to use script target when required
* remove invisibility aura (18) based on attributes. Passive and negative auras will
  not be removed on combat
* implement spell effect 19395
* implement Battleground scores storage system
* implement EVENT_T_ENERGY for EventAI
* SCRIPT_COMMAND_CAST_SPELL (15) allows passing additional spell-IDs with
  dataint..dataint4 columns. If these are used, a random spell out of datalong,
  dataint, ..dataintX is used as cast spell.
* Lua 5.2 is now an external dependency. Please install Lua 5.2 with development
  headers. Windows users should grab [Lua binaries](http://luabinaries.sourceforge.net/)
  and Linux users will find a Lua 5.2 package from their distribution vendor, e.g.
  Debian/Ubuntu provide `liblua5.2-0`/`liblua5.2-dev` packages.

Minor fixes and improvements have been added, such as:

* the CMake build system is now compatible with CMake 2.8 and CMake 3.
* fixed the range of traps based on YouTube videos.
* fixed the range in which spell focus objects (such as fires) will trigger.
  This prevents fire damage taken even though you're not standing in the fire.
* database files and updates will now be installed, too.
* code refactoring in many functions makes the code more readable.
* rooted players will now not only appear rooted to other players, but actually
  stay rooted for real.
* fixed skinning loot bug.
* allow non-instanced empty map creation.
* replaced `Map::LoadGrid` with `Map::ForceLoadGrid`
* fixed raid instance reset crash
* use strict database filter for creatures with movement points without a
  matching movement type
* fix possible mount abuse after leaving battlegrounds
* unify startup messages, and changed loading routines for sane
  grouping
* use a global team index for battleground teams
* fixed a bug that caused AreaAura reapply because the code doesn't search the
  correct rank of it.
* fixed AreaAura bug when low level grouped with high level Player. Thank to
  @tarwyn for his precious help and proposed patch.
* allow logout during duels
* various code style fixes
* various performance fixes
* CreatureLinking: improve startup efficiency. No need to use multiple select to
  check and obtain information about db-data of an expectedly unique creature
* CreatureLinking: allow recursive CanSpawn checks. This will fix behaviour of
  mobs whose spawning is dependent on mobs whose spawning is dependent on other
  mobs.

## 1.0.0 (2014-10-27)

Minor fixes and improvements have been added, such as:

* usernames are now normalized when updating the account password, no longer
  resulting in the creation of a broken SHA1 hash due to lowercase characters
  in usernames.
* an unused setting for max character level has been removed.
* paths will no longer be recalculated on movement speed changing effects on
  creatures. This removes stuttering when creatures are e.g. hit by Frost Bolt.
* removed support for [Shiv](http://wowpedia.org/Shiv) which was introduced
  in The Burning Crusade first.
* grouped players will now see stealthed members with the proper animation.
* fixed build for Visual Studio 2012 (/bigobj required).
* the build system is now fully compatible with CMake 3.0, and less spammy
  when building Lua.

## 1.0.0-RC2 (2014-10-16)

Major changes for this build which require your attention when upgrading include
things such as these:

* removal of PostgreSQL support. Sadly, due to lack of testing and development
  resources required to maintain proper support, we had to remove support for
  PostgreSQL.
* removal of the auction house bot. Sadly, due to lack of testing and development

src/game/Item.h
/**
 * mangos-zero is a full featured server for World of Warcraft in its vanilla
 * version, supporting clients for patch 1.12.x.
 *
 * Copyright (C) 2005-2014  MaNGOS project  <http://getmangos.com>
 * Parts Copyright (C) 2013-2014  CMaNGOS project <http://cmangos.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * World of Warcraft, and all World of Warcraft or Warcraft art, images,
 * and lore are copyrighted by Blizzard Entertainment, Inc.
 */

#ifndef MANGOS_H_ITEM
#define MANGOS_H_ITEM

#include "Common.h"
#include "Object.h"
#include "LootMgr.h"
#include "ItemPrototype.h"

struct SpellEntry;
class Bag;
class Field;
class QueryResult;
class Unit;

struct ItemSetEffect
{
    uint32 setid;
    uint32 item_count;
    SpellEntry const* spells[8];
};

// [-ZERO] Need fix, possible uptodate in mangos-0.6
enum InventoryResult
{
    EQUIP_ERR_OK                                 = 0,
    EQUIP_ERR_CANT_EQUIP_LEVEL_I                 = 1,       // ERR_CANT_EQUIP_LEVEL_I
    EQUIP_ERR_CANT_EQUIP_SKILL                   = 2,       // ERR_CANT_EQUIP_SKILL
    EQUIP_ERR_ITEM_DOESNT_GO_TO_SLOT             = 3,       // ERR_WRONG_SLOT
    EQUIP_ERR_BAG_FULL                           = 4,       // ERR_BAG_FULL
    EQUIP_ERR_NONEMPTY_BAG_OVER_OTHER_BAG        = 5,       // ERR_BAG_IN_BAG
    EQUIP_ERR_CANT_TRADE_EQUIP_BAGS              = 6,       // ERR_TRADE_EQUIPPED_BAG
    EQUIP_ERR_ONLY_AMMO_CAN_GO_HERE              = 7,       // ERR_AMMO_ONLY
    EQUIP_ERR_NO_REQUIRED_PROFICIENCY            = 8,       // ERR_PROFICIENCY_NEEDED
    EQUIP_ERR_NO_EQUIPMENT_SLOT_AVAILABLE        = 9,       // ERR_NO_SLOT_AVAILABLE
    EQUIP_ERR_YOU_CAN_NEVER_USE_THAT_ITEM        = 10,      // ERR_CANT_EQUIP_EVER
    EQUIP_ERR_YOU_CAN_NEVER_USE_THAT_ITEM2       = 11,      // ERR_CANT_EQUIP_EVER
    EQUIP_ERR_NO_EQUIPMENT_SLOT_AVAILABLE2       = 12,      // ERR_NO_SLOT_AVAILABLE
    EQUIP_ERR_CANT_EQUIP_WITH_TWOHANDED          = 13,      // ERR_2HANDED_EQUIPPED
    EQUIP_ERR_CANT_DUAL_WIELD                    = 14,      // ERR_2HSKILLNOTFOUND
    EQUIP_ERR_ITEM_DOESNT_GO_INTO_BAG            = 15,      // ERR_WRONG_BAG_TYPE
    EQUIP_ERR_ITEM_DOESNT_GO_INTO_BAG2           = 16,      // ERR_WRONG_BAG_TYPE
    EQUIP_ERR_CANT_CARRY_MORE_OF_THIS            = 17,      // ERR_ITEM_MAX_COUNT
    EQUIP_ERR_NO_EQUIPMENT_SLOT_AVAILABLE3       = 18,      // ERR_NO_SLOT_AVAILABLE
    EQUIP_ERR_ITEM_CANT_STACK                    = 19,      // ERR_CANT_STACK
    EQUIP_ERR_ITEM_CANT_BE_EQUIPPED              = 20,      // ERR_NOT_EQUIPPABLE
    EQUIP_ERR_ITEMS_CANT_BE_SWAPPED              = 21,      // ERR_CANT_SWAP
    EQUIP_ERR_SLOT_IS_EMPTY                      = 22,      // ERR_SLOT_EMPTY
    EQUIP_ERR_ITEM_NOT_FOUND                     = 23,      // ERR_ITEM_NOT_FOUND
    EQUIP_ERR_CANT_DROP_SOULBOUND                = 24,      // ERR_DROP_BOUND_ITEM
    EQUIP_ERR_OUT_OF_RANGE                       = 25,      // ERR_OUT_OF_RANGE
    EQUIP_ERR_TRIED_TO_SPLIT_MORE_THAN_COUNT     = 26,      // ERR_TOO_FEW_TO_SPLIT
    EQUIP_ERR_COULDNT_SPLIT_ITEMS                = 27,      // ERR_SPLIT_FAILED
    EQUIP_ERR_MISSING_REAGENT                    = 28,      // ERR_SPELL_FAILED_REAGENTS_GENERIC
    EQUIP_ERR_NOT_ENOUGH_MONEY                   = 29,      // ERR_NOT_ENOUGH_MONEY
    EQUIP_ERR_NOT_A_BAG                          = 30,      // ERR_NOT_A_BAG
    EQUIP_ERR_CAN_ONLY_DO_WITH_EMPTY_BAGS        = 31,      // ERR_DESTROY_NONEMPTY_BAG
    EQUIP_ERR_DONT_OWN_THAT_ITEM                 = 32,      // ERR_NOT_OWNER
    EQUIP_ERR_CAN_EQUIP_ONLY1_QUIVER             = 33,      // ERR_ONLY_ONE_QUIVER
    EQUIP_ERR_MUST_PURCHASE_THAT_BAG_SLOT        = 34,      // ERR_NO_BANK_SLOT
    EQUIP_ERR_TOO_FAR_AWAY_FROM_BANK             = 35,      // ERR_NO_BANK_HERE
    EQUIP_ERR_ITEM_LOCKED                        = 36,      // ERR_ITEM_LOCKED
    EQUIP_ERR_YOU_ARE_STUNNED                    = 37,      // ERR_GENERIC_STUNNED
    EQUIP_ERR_YOU_ARE_DEAD                       = 38,      // ERR_PLAYER_DEAD
    EQUIP_ERR_CANT_DO_RIGHT_NOW                  = 39,      // ERR_CLIENT_LOCKED_OUT
    EQUIP_ERR_INT_BAG_ERROR                      = 40,      // ERR_INTERNAL_BAG_ERROR
    EQUIP_ERR_CAN_EQUIP_ONLY1_BOLT               = 41,      // ERR_ONLY_ONE_BOLT
    EQUIP_ERR_CAN_EQUIP_ONLY1_AMMOPOUCH          = 42,      // ERR_ONLY_ONE_AMMO
    EQUIP_ERR_STACKABLE_CANT_BE_WRAPPED          = 43,      // ERR_CANT_WRAP_STACKABLE
    EQUIP_ERR_EQUIPPED_CANT_BE_WRAPPED           = 44,      // ERR_CANT_WRAP_EQUIPPED
    EQUIP_ERR_WRAPPED_CANT_BE_WRAPPED            = 45,      // ERR_CANT_WRAP_WRAPPED
    EQUIP_ERR_BOUND_CANT_BE_WRAPPED              = 46,      // ERR_CANT_WRAP_BOUND
    EQUIP_ERR_UNIQUE_CANT_BE_WRAPPED             = 47,      // ERR_CANT_WRAP_UNIQUE
    EQUIP_ERR_BAGS_CANT_BE_WRAPPED               = 48,      // ERR_CANT_WRAP_BAGS
    EQUIP_ERR_ALREADY_LOOTED                     = 49,      // ERR_LOOT_GONE
    EQUIP_ERR_INVENTORY_FULL                     = 50,      // ERR_INV_FULL
    EQUIP_ERR_BANK_FULL                          = 51,      // ERR_BAG_FULL
    EQUIP_ERR_ITEM_IS_CURRENTLY_SOLD_OUT         = 52,      // ERR_VENDOR_SOLD_OUT
    EQUIP_ERR_BAG_FULL3                          = 53,      // ERR_BAG_FULL
    EQUIP_ERR_ITEM_NOT_FOUND2                    = 54,      // ERR_ITEM_NOT_FOUND
    EQUIP_ERR_ITEM_CANT_STACK2                   = 55,      // ERR_CANT_STACK
    EQUIP_ERR_BAG_FULL4                          = 56,      // ERR_BAG_FULL
    EQUIP_ERR_ITEM_SOLD_OUT                      = 57,      // ERR_VENDOR_SOLD_OUT
    EQUIP_ERR_OBJECT_IS_BUSY                     = 58,      // ERR_OBJECT_IS_BUSY
    EQUIP_ERR_NONE                               = 59,      // ERR_CANT_BE_DISENCHANTED
    EQUIP_ERR_NOT_IN_COMBAT                      = 60,      // ERR_NOT_IN_COMBAT
    EQUIP_ERR_NOT_WHILE_DISARMED                 = 61,      // ERR_NOT_WHILE_DISARMED
    EQUIP_ERR_BAG_FULL6                          = 62,      // ERR_BAG_FULL
    EQUIP_ERR_CANT_EQUIP_RANK                    = 63,      // ERR_CANT_EQUIP_RANK
    EQUIP_ERR_CANT_EQUIP_REPUTATION              = 64,      // ERR_CANT_EQUIP_REPUTATION
    EQUIP_ERR_TOO_MANY_SPECIAL_BAGS              = 65,      // ERR_TOO_MANY_SPECIAL_BAGS
    EQUIP_ERR_LOOT_CANT_LOOT_THAT_NOW            = 66,      // ERR_LOOT_CANT_LOOT_THAT_NOW
    // any greater values show as "bag full"
};

enum BuyResult
{
    BUY_ERR_CANT_FIND_ITEM                      = 0,
    BUY_ERR_ITEM_ALREADY_SOLD                   = 1,
    BUY_ERR_NOT_ENOUGHT_MONEY                   = 2,
    BUY_ERR_SELLER_DONT_LIKE_YOU                = 4,
    BUY_ERR_DISTANCE_TOO_FAR                    = 5,
    BUY_ERR_ITEM_SOLD_OUT                       = 7,
    BUY_ERR_CANT_CARRY_MORE                     = 8,
    BUY_ERR_RANK_REQUIRE                        = 11,
    BUY_ERR_REPUTATION_REQUIRE                  = 12
};

enum SellResult
{
    SELL_ERR_CANT_FIND_ITEM                      = 1,
    SELL_ERR_CANT_SELL_ITEM                      = 2,       // merchant doesn't like that item
    SELL_ERR_CANT_FIND_VENDOR                    = 3,       // merchant doesn't like you
    SELL_ERR_YOU_DONT_OWN_THAT_ITEM              = 4,       // you don't own that item
    SELL_ERR_UNK                                 = 5,       // nothing appears...
    SELL_ERR_ONLY_EMPTY_BAG                      = 6        // can only do with empty bags
};

// -1 from client enchantment slot number
enum EnchantmentSlot
{
    PERM_ENCHANTMENT_SLOT       = 0,
    TEMP_ENCHANTMENT_SLOT       = 1,
    MAX_INSPECTED_ENCHANTMENT_SLOT = 2,

    PROP_ENCHANTMENT_SLOT_0     = 3,                        // used with RandomSuffix
    PROP_ENCHANTMENT_SLOT_1     = 4,                        // used with RandomSuffix
    PROP_ENCHANTMENT_SLOT_2     = 5,                        // used with RandomSuffix
    PROP_ENCHANTMENT_SLOT_3     = 6,
    MAX_ENCHANTMENT_SLOT        = 7
};

#define MAX_VISIBLE_ITEM_OFFSET       12

enum EnchantmentOffset
{
    ENCHANTMENT_ID_OFFSET       = 0,
    ENCHANTMENT_DURATION_OFFSET = 1,
    ENCHANTMENT_CHARGES_OFFSET  = 2
};

#define MAX_ENCHANTMENT_OFFSET    3

enum EnchantmentSlotMask
{
    ENCHANTMENT_CAN_SOULBOUND  = 0x01,
    ENCHANTMENT_UNK1           = 0x02,
    ENCHANTMENT_UNK2           = 0x04,
    ENCHANTMENT_UNK3           = 0x08
};

enum ItemUpdateState
{
    ITEM_UNCHANGED                               = 0,
    ITEM_CHANGED                                 = 1,
    ITEM_NEW                                     = 2,
    ITEM_REMOVED                                 = 3
};

enum ItemLootUpdateState
{
    ITEM_LOOT_NONE                                = 0,      // loot not generated
    ITEM_LOOT_TEMPORARY                           = 1,      // generated loot is temporary (will deleted at loot window close)
    ITEM_LOOT_UNCHANGED                           = 2,
    ITEM_LOOT_CHANGED                             = 3,
    ITEM_LOOT_NEW                                 = 4,
    ITEM_LOOT_REMOVED                             = 5
};

// masks for ITEM_FIELD_FLAGS field
enum ItemDynFlags
{
    ITEM_DYNFLAG_BINDED                       = 0x00000001, // set in game at binding
    ITEM_DYNFLAG_UNK1                         = 0x00000002,
    ITEM_DYNFLAG_UNLOCKED                     = 0x00000004, // have meaning only for item with proto->LockId, if not set show as "Locked, req. lockpicking N"
    ITEM_DYNFLAG_WRAPPED                      = 0x00000008, // mark item as wrapped into wrapper container
    ITEM_DYNFLAG_UNK4                         = 0x00000010, // can't repeat old note: appears red icon (like when item durability==0)
    ITEM_DYNFLAG_UNK5                         = 0x00000020,
    ITEM_DYNFLAG_UNK6                         = 0x00000040, // ? old note: usable
    ITEM_DYNFLAG_UNK7                         = 0x00000080,
    ITEM_DYNFLAG_UNK8                         = 0x00000100,
    ITEM_DYNFLAG_READABLE                     = 0x00000200, // can be open for read, it or item proto pagetText make show "Right click to read"
    ITEM_DYNFLAG_UNK10                        = 0x00000400,
    ITEM_DYNFLAG_UNK11                        = 0x00000800,
    ITEM_DYNFLAG_UNK12                        = 0x00001000,
    ITEM_DYNFLAG_UNK13                        = 0x00002000,
    ITEM_DYNFLAG_UNK14                        = 0x00004000,
    ITEM_DYNFLAG_UNK15                        = 0x00008000,
    ITEM_DYNFLAG_UNK16                        = 0x00010000,
    ITEM_DYNFLAG_UNK17                        = 0x00020000,
};

enum ItemRequiredTargetType
{
    ITEM_TARGET_TYPE_CREATURE   = 1,
    ITEM_TARGET_TYPE_DEAD       = 2
};

#define MAX_ITEM_REQ_TARGET_TYPE 2

struct ItemRequiredTarget
{
    ItemRequiredTarget(ItemRequiredTargetType uiType, uint32 uiTargetEntry) : m_uiType(uiType), m_uiTargetEntry(uiTargetEntry) {}
    ItemRequiredTargetType m_uiType;
    uint32 m_uiTargetEntry;

    // helpers
    bool IsFitToRequirements(Unit* pUnitTarget) const;
};

bool ItemCanGoIntoBag(ItemPrototype const* proto, ItemPrototype const* pBagProto);

class MANGOS_DLL_SPEC Item : public Object
{
    public:
        static Item* CreateItem(uint32 item, uint32 count, Player const* player = NULL, uint32 randomPropertyId = 0);
        Item* CloneItem(uint32 count, Player const* player = NULL) const;

        Item();
        ~Item();

        virtual bool Create(uint32 guidlow, uint32 itemid, Player const* owner);

        ItemPrototype const* GetProto() const;

        ObjectGuid const& GetOwnerGuid() const
        {
            return GetGuidValue(ITEM_FIELD_OWNER);
        }
        void SetOwnerGuid(ObjectGuid guid)
        {
            SetGuidValue(ITEM_FIELD_OWNER, guid);
        }
        Player* GetOwner()const;

        void SetBinding(bool val)
        {
            ApplyModFlag(ITEM_FIELD_FLAGS, ITEM_DYNFLAG_BINDED, val);
        }
        bool IsSoulBound() const
        {
            return HasFlag(ITEM_FIELD_FLAGS, ITEM_DYNFLAG_BINDED);
        }
        bool IsBindedNotWith(Player const* player) const;
        bool IsBoundByEnchant() const;
        virtual void SaveToDB();
        virtual bool LoadFromDB(uint32 guidLow, Field* fields, ObjectGuid ownerGuid = ObjectGuid());
        virtual void DeleteFromDB();
        void DeleteFromInventoryDB();
        void LoadLootFromDB(Field* fields);

        Bag* ToBag()
        {
            if (IsBag()) return reinterpret_cast<Bag*>(this);
            else return NULL;
        }
        const Bag* ToBag() const
        {
            if (IsBag()) return reinterpret_cast<const Bag*>(this);
            else return NULL;
        }

        bool IsLocked() const
        {
            return !HasFlag(ITEM_FIELD_FLAGS, ITEM_DYNFLAG_UNLOCKED);
        }
        bool IsBag() const
        {
            return GetProto()->InventoryType == INVTYPE_BAG;
        }
        bool IsNotEmptyBag() const;
        bool IsBroken() const
        {
            return GetUInt32Value(ITEM_FIELD_MAXDURABILITY) > 0 && GetUInt32Value(ITEM_FIELD_DURABILITY) == 0;
        }
        bool CanBeTraded() const;
        void SetInTrade(bool b = true)
        {
            mb_in_trade = b;
        }
        bool IsInTrade() const
        {
            return mb_in_trade;
        }

        bool IsFitToSpellRequirements(SpellEntry const* spellInfo) const;
        bool IsTargetValidForItemUse(Unit* pUnitTarget);
        bool IsLimitedToAnotherMapOrZone(uint32 cur_mapId, uint32 cur_zoneId) const;

        uint32 GetCount() const
        {
            return GetUInt32Value(ITEM_FIELD_STACK_COUNT);
        }
        void SetCount(uint32 value)
        {
            SetUInt32Value(ITEM_FIELD_STACK_COUNT, value);
        }
        uint32 GetMaxStackCount() const
        {
            return GetProto()->GetMaxStackSize();
        }
        InventoryResult CanBeMergedPartlyWith(ItemPrototype const* proto) const;

        uint8 GetSlot() const
        {
            return m_slot;
        }
        Bag* GetContainer()
        {
            return m_container;
        }
        uint8 GetBagSlot() const;
        void SetSlot(uint8 slot)
        {
            m_slot = slot;
        }
        uint16 GetPos() const
        {
            return uint16(GetBagSlot()) << 8 | GetSlot();
        }
        void SetContainer(Bag* container)
        {
            m_container = container;
        }

        bool IsInBag() const
        {
            return m_container != NULL;
        }
        bool IsEquipped() const;

        uint32 GetSkill();
        uint32 GetSpell();

        // RandomPropertyId (signed but stored as unsigned)
        int32 GetItemRandomPropertyId() const
        {
            return GetInt32Value(ITEM_FIELD_RANDOM_PROPERTIES_ID);
        }
        uint32 GetItemSuffixFactor() const
        {
            return GetUInt32Value(ITEM_FIELD_PROPERTY_SEED);
        }
        void SetItemRandomProperties(int32 randomPropId);
        static int32 GenerateItemRandomPropertyId(uint32 item_id);
        void SetEnchantment(EnchantmentSlot slot, uint32 id, uint32 duration, uint32 charges);
        void SetEnchantmentDuration(EnchantmentSlot slot, uint32 duration);
        void SetEnchantmentCharges(EnchantmentSlot slot, uint32 charges);
        void ClearEnchantment(EnchantmentSlot slot);
        uint32 GetEnchantmentId(EnchantmentSlot slot)       const
        {
            return GetUInt32Value(ITEM_FIELD_ENCHANTMENT + slot * MAX_ENCHANTMENT_OFFSET + ENCHANTMENT_ID_OFFSET);
        }
        uint32 GetEnchantmentDuration(EnchantmentSlot slot) const
        {
            return GetUInt32Value(ITEM_FIELD_ENCHANTMENT + slot * MAX_ENCHANTMENT_OFFSET + ENCHANTMENT_DURATION_OFFSET);
        }
        uint32 GetEnchantmentCharges(EnchantmentSlot slot)  const
        {
            return GetUInt32Value(ITEM_FIELD_ENCHANTMENT + slot * MAX_ENCHANTMENT_OFFSET + ENCHANTMENT_CHARGES_OFFSET);
        }

        void SendTimeUpdate(Player* owner);
        void UpdateDuration(Player* owner, uint32 diff);

        // spell charges (signed but stored as unsigned)
        int32 GetSpellCharges(uint8 index/*0..5*/ = 0) const
        {
            return GetInt32Value(ITEM_FIELD_SPELL_CHARGES + index);
        }
        void SetSpellCharges(uint8 index/*0..5*/, int32 value)
        {
            SetInt32Value(ITEM_FIELD_SPELL_CHARGES + index, value);
        }

        Loot loot;

        void SetLootState(ItemLootUpdateState state);
        bool HasGeneratedLoot() const
        {
            return m_lootState != ITEM_LOOT_NONE && m_lootState != ITEM_LOOT_REMOVED;
        }
        bool HasTemporaryLoot() const
        {
            return m_lootState == ITEM_LOOT_TEMPORARY;
        }

        bool HasSavedLoot() const
        {
            return m_lootState != ITEM_LOOT_NONE && m_lootState != ITEM_LOOT_NEW && m_lootState != ITEM_LOOT_TEMPORARY;
        }

        // Update States
        ItemUpdateState GetState() const
        {
            return uState;
        }
        void SetState(ItemUpdateState state, Player* forplayer = NULL);
        void AddToUpdateQueueOf(Player* player);
        void RemoveFromUpdateQueueOf(Player* player);
        bool IsInUpdateQueue() const
        {
            return uQueuePos != -1;
        }
        uint16 GetQueuePos() const
        {
            return uQueuePos;
        }
        void FSetState(ItemUpdateState state)               // forced
        {
            uState = state;
        }

        bool HasQuest(uint32 quest_id) const override
        {
            return GetProto()->StartQuest == quest_id;
        }
        bool HasInvolvedQuest(uint32 /*quest_id*/) const override
        {
            return false;
        }
        bool IsPotion() const
        {
            return GetProto()->IsPotion();
        }
        bool IsConjuredConsumable() const
        {
            return GetProto()->IsConjuredConsumable();
        }
        bool IsWeaponVellum() const
        {
            return GetProto()->IsWeaponVellum();
        }
        bool IsArmorVellum() const
        {
            return GetProto()->IsArmorVellum();
        }

        void AddToClientUpdateList() override;
        void RemoveFromClientUpdateList() override;
        void BuildUpdateData(UpdateDataMapType& update_players) override;
    private:
        uint8 m_slot;
        Bag* m_container;
        ItemUpdateState uState;
        int16 uQueuePos;
        bool mb_in_trade;                                   // true if item is currently in trade-window
        ItemLootUpdateState m_lootState;
};

#endif
